File: vpid/vpi/src/uploader.rs
//! Uploader submodule to upload new firmware using custom stm8sboot loader.
//! Boot loader code is here: https://github.com/ludiazv/stm8-bootloader
//!
use i2cdev::core::I2CDevice;
use i2cdev::linux::*;
use pbr::ProgressBar;
use std::fs::OpenOptions;
use std::io::ErrorKind;
use std::io::Read;
use std::path::PathBuf;
use std::thread::sleep;
use std::time::Duration;
use sysfs_gpio::{Direction, Pin};

/// BLOCK SIZE IS 64 for stms8 low desity
const BLOCK_SIZE: usize = 64;
const ACK: [u8; 2] = [0xaa, 0xbb];
const NACK: [u8; 2] = [0xde, 0xad];

/// Just a simple crc8 computation compatble with bootloader
fn crc8_update(data: u8, crc_in: u8) -> u8 {
    let mut crc: u8 = crc_in ^ data;
    for _i in 0..8 {
        if crc & 0x80 != 0 {
            crc = (crc << 1) ^ 0x07;
        } else {
            crc <<= 1;
        }
    }
    return crc & 0xFF;
}

/// Compute crc8 of a buffer
pub fn buff_crc(data: &[u8], mut crc_in: u8) -> u8 {
    for i in data.iter() {
        crc_in = crc8_update(*i, crc_in);
    }
    return crc_in;
}

/// Compute the crc of a binary file
/// # Arguments
/// `file` : Path of the file
///
///
fn get_crc(file: &PathBuf) -> Result<(u8, u8), std::io::Error> {
    let mut crc: u8 = 0;
    let mut chunk: [u8; BLOCK_SIZE] = [0xFF; BLOCK_SIZE];
    let mut f = OpenOptions::new().read(true).open(file)?;
    let size = f.metadata()?.len();
    let mut sbl = size / (BLOCK_SIZE as u64);
    sbl = if (size % (BLOCK_SIZE as u64)) > 0 {
        sbl + 1
    } else {
        sbl
    };
    let mut len = f.read(&mut chunk)?;
    while len > 0 {
        crc = buff_crc(&chunk, crc);
        chunk = [0xFF; BLOCK_SIZE]; // Reset chuck
        len = f.read(&mut chunk)?; // Read next chunck
    }
    Ok((crc, sbl as u8))
}

/// Resets the STM8S with High pulse using a pin of the SBC
/// # Arguments
/// `pin`- Pin number
fn reset(pin: u16) -> Result<(), std::io::Error> {
    let pin = Pin::new(pin as u64);

    let block = || -> Result<(), sysfs_gpio::Error> {
        pin.export()?;
        sleep(Duration::from_millis(3500)); // Big delay for export
        pin.set_direction(Direction::Out)?;
        pin.set_value(0)?;
        sleep(Duration::from_millis(500));
        pin.set_value(1)?;
        sleep(Duration::from_millis(500));
        pin.set_value(0)?;
        Ok(())
    };

    match block() {
        Ok(()) => Ok(()),
        Err(e) => {
            println!("{:?}", e);
            Err(std::io::Error::new(
                ErrorKind::PermissionDenied,
                "could not reset vpi board via gpio, please check conection & permissions",
            ))
        }
    }
}

/// Unexport a pin
/// # Arguments
/// `pin` - pin to unexport
pub fn unexport(pin: u16) {
    let pin = Pin::new(pin as u64);
    let _ = pin.unexport();
}

/// Small pause
#[inline]
fn pause() {
    sleep(Duration::from_millis(20));
}
/// Uploads the firmeware using i2c
/// # Arguments
/// `addr` - i2c address of the i2c board
/// `dev_path` - path to i2c devive /dev/i2c-XX
/// `file` - file path of the firmware
/// `rst_pin` - pin used for reset the stm8s chip (rest will be high level)
pub fn upload(
    addr: u8,
    dev_path: &PathBuf,
    file: &PathBuf,
    rst_pin: u16,
) -> Result<(), LinuxI2CError> {
    let mut dev = LinuxI2CDevice::new(dev_path, addr as u16)?;
    let crc: u8;
    let blocks: u8;
    if file.exists() {
        // Compute CRC
        match get_crc(file) {
            Ok((c, l)) => {
                crc = c;
                blocks = l;
            }
            Err(e) => return Err(LinuxI2CError::Io(e)),
        }
        println!("Firmware readed. blocks={},CRC={:x}", blocks, crc);
        // Initiate bootloader sequence
        reset(rst_pin)?;
        let req: [u8; 7] = [0xde, 0xad, 0xbe, 0xef, blocks, crc, crc]; // Activation msg
        let mut resp: [u8; 2] = NACK;
        print!("Sending upload request...");
        pause();
        dev.write(&req)?;
        pause();
        print!("R.");
        dev.read(&mut resp)?;
        if resp != ACK {
            return Err(LinuxI2CError::Io(std::io::Error::new(
                ErrorKind::ConnectionRefused,
                "Bootloader activation response:NACK",
            )));
        }
        println!("Ok! Starting upload...");
        // Upload chucks
        let mut pb = ProgressBar::new(blocks as u64);
        let mut f = OpenOptions::new().read(true).open(file)?;
        for _i in 0..blocks {
            let mut chunk: [u8; BLOCK_SIZE] = [0xFF; BLOCK_SIZE];
            f.read(&mut chunk)?;
            dev.write(&chunk)?;
            pb.inc();
            pause();
        }
        // ACK confirmation of the firmware
        resp = NACK;
        print!("Confirming upload...");
        pause();
        dev.read(&mut resp)?;
        if resp != ACK {
            Err(LinuxI2CError::Io(std::io::Error::new(
                ErrorKind::ConnectionRefused,
                "Bootloader activation response:NACK",
            )))
        } else {
            println!("Ok!");
            Ok(())
        }
    } else {
        Err(LinuxI2CError::Io(std::io::Error::new(
            ErrorKind::NotFound,
            "Firmware file not found",
        )))
    }
}

File: firmware_borosVPi/misc/uploader.py
# STM8S uploader module for micropython
import os
import time
import math

BLOCK_SIZE  = const(64)
BOOT_ADDR   = const(0x22)
BOOT_ACK    = b'\xaa\xbb' 


def crc8_update(data, crc):
    crc ^= data
    for i in range(0, 8):
        if crc & 0x80 != 0:
            crc = (crc << 1) ^ 0x07
        else:
            crc <<= 1
    return crc & 0xFF

def get_crc(file):
    crc = 0
    data = open(file, 'rb')
    with data as f:
        chunk = f.read(BLOCK_SIZE)
        while chunk:
            chunk = chunk + (b'\xFF' * (BLOCK_SIZE - len(chunk)))
            for i in chunk:
                crc = crc8_update(i, crc)
            chunk = f.read(BLOCK_SIZE)
    return crc.to_bytes(1,"big")


class Stm8s_i2c_uploader:
    def __init__(self,i2c,rst_pin,inverted=True):
        self.i2c=i2c
        self.rst_pin=rst_pin
        if inverted:
            self.high=0
            self.low=1
        else:
            self.high=1
            self.low=0
        self.crc=0
        
    def rst(self):
        self.rst_pin.value(self.high)
        time.sleep_ms(100)
        self.rst_pin.value(self.low)
        time.sleep_ms(300)
        self.rst_pin.value(self.high)
        time.sleep_ms(15)
    
    def enter_bootloader(self,file):
        req= b'\xde\xad\xbe\xef'
        n=math.ceil(os.stat(file)[6]/BLOCK_SIZE)
        req= req + n.to_bytes(1,"big") + self.crc + self.crc
        self.rst()
        self.i2c.writeto(BOOT_ADDR,req)
        time.sleep_ms(10)
        req=self.i2c.readfrom(BOOT_ADDR,2)
        return (req == BOOT_ACK)


    def upload(self,file):
        # Compute the CRC of file
        self.crc=get_crc(file)
        # Enter into booot loader
        if self.enter_bootloader(file):
            data = open(file, 'rb')
            #total = 0
            #idx = 0
            with data as f:
                chunk = f.read(BLOCK_SIZE)
                while chunk:
                    chunk=chunk  + (b'\xFF' * (BLOCK_SIZE - len(chunk)))
                    self.i2c.writeto(BOOT_ADDR,chunk)                    
                    chunk = f.read(BLOCK_SIZE)
                    time.sleep_ms(50) # I2c could be to fast , give some time to process the block
        
            time.sleep_ms(10)
            ret=self.i2c.readfrom(BOOT_ADDR,2)
            self.rst()
            return (ret == BOOT_ACK)
        else:
            return False







